Daemon segfault

**Describe the bug**

My intercept stopped working without anything obvious popping up. When I was trying to restart the intercept I got this:

```
telepresence intercept: error: connection error: desc = "transport: error while dialing: dial unix /var/run/telepresence-daemon.socket: connect: connection refused" (socket rm failed with remove /var/run/telepresence-daemon.socket: permission denied); this usually means that the process has terminated ungracefully
```

This is the dump from the daemon log:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x17dfa3b]

goroutine 66993 [running]:
github.com/telepresenceio/telepresence/v2/pkg/tunnel.readLoop({0x2629b68, 0xc000b31e30}, {0x262f648, 0xc000858e70})
	github.com/telepresenceio/telepresence/v2/pkg/tunnel/dialer.go:294 +0x3db
github.com/telepresenceio/telepresence/v2/pkg/tunnel.(*dialer).streamToConnLoop(0x0?, {0x2629b68?, 0xc000b31e30?}, 0x0?)
	github.com/telepresenceio/telepresence/v2/pkg/tunnel/dialer.go:239 +0x6e
created by github.com/telepresenceio/telepresence/v2/pkg/tunnel.(*dialer).Start.func1
	github.com/telepresenceio/telepresence/v2/pkg/tunnel/dialer.go:143 +0xaaf
```

**Versions (please complete the following information):**

```
$ telepresence version
Client         : v2.9.5
Root Daemon    : v2.9.5
User Daemon    : v2.9.5
```

--- a/.github/actions/integration-tests/action.yaml
+++ b/.github/actions/integration-tests/action.yaml
@@ -11,6 +11,7 @@ runs:
       env:
         DTEST_KUBECONFIG: "${{ inputs.kubeconfig }}"
         DTEST_REGISTRY: "docker.io/datawire"
+        SCOUT_DISABLE: "1"
       uses: nick-invision/retry@v2.8.2
       with:
         max_attempts: 2
@@ -19,15 +20,7 @@ runs:
         command: |
           set -ex
           if [[ ${RUNNER_OS} == "Windows" ]]; then
-            # We want to validate that tests still pass, even if the metrics host
-            # points to a broken IP
-            echo "127.0.0.1 metriton.datawire.io" >> c:\windows\system32\drivers\etc\hosts
-
             export PATH="$PATH:/C/Program Files/SSHFS-Win/bin"
-          else
-            # We want to validate that tests still pass, even if the metrics host
-            # points to a broken IP
-            echo "127.0.0.1 metriton.datawire.io" | sudo tee -a /etc/hosts
           fi
 
           make check-integration

--- a/CHANGELOG.yml
+++ b/CHANGELOG.yml
@@ -54,6 +54,13 @@ items:
           A traffic-agent of version 2.13.3 (or 1.13.15) would not propagate the directories under
           <code>/var/run/secrets</code> when used with a traffic manager older than 2.13.3.
 
+      - type: bugfix
+        title: Fixed race condition causing segfaults on rare occasions when a tunnel stream timed out.
+        body: >-
+          A context cancellation could sometimes be trapped in a stream reader, causing it to incorrectly return
+          an undefined message which in turn caused the parent reader to panic on a <code>nil</code> pointer reference.
+        docs: https://github.com/telepresenceio/telepresence/pull/2963
+
       - type: change
         title: Routing conflict reporting.
         body: >-

--- a/integration_test/headless_test.go
+++ b/integration_test/headless_test.go
@@ -2,7 +2,6 @@ package integration_test
 
 import (
 	"context"
-	"fmt"
 	"strconv"
 	"strings"
 	"time"
@@ -24,107 +23,21 @@ func (s *connectedSuite) Test_SuccessfullyInterceptsHeadlessService() {
 
 	s.ApplyApp(ctx, "echo-headless", "statefulset/echo-headless")
 	defer s.DeleteSvcAndWorkload(ctx, "statefulset", "echo-headless")
+	require := s.Require()
+	stdout := itest.TelepresenceOk(ctx, "intercept", "--namespace", s.AppNamespace(), "--mount", "false", svc, "--port", strconv.Itoa(svcPort))
+	require.Contains(stdout, "Using StatefulSet echo-headless")
+	s.CapturePodLogs(ctx, "service=echo-headless", "traffic-agent", s.AppNamespace())
 
-	for _, test := range []struct {
-		webhook bool
-		name    string
-	}{
-		{
-			webhook: true,
-			name:    "injected from webhook",
+	defer itest.TelepresenceOk(ctx, "leave", "echo-headless-"+s.AppNamespace())
+
+	require.Eventually(
+		func() bool {
+			stdout, _, err := itest.Telepresence(ctx, "list", "--namespace", s.AppNamespace(), "--intercepts")
+			return err == nil && strings.Contains(stdout, "echo-headless: intercepted")
 		},
-		{
-			webhook: false,
-			name:    "injected from command",
-		},
-	} {
-		s.Run(test.name, func() {
-			require := s.Require()
-			ctx := s.Context()
-			if test.webhook {
-				require.NoError(annotateForWebhook(ctx, "statefulset", "echo-headless", s.AppNamespace(), 8080))
-				require.Eventually(
-					func() bool {
-						stdout, _, err := itest.Telepresence(ctx, "list", "--namespace", s.AppNamespace(), "--agents")
-						return err == nil && strings.Contains(stdout, "echo-headless: ready to intercept")
-					},
-					30*time.Second, // waitFor
-					3*time.Second,  // polling interval
-					`never gets install agent`)
-			}
-			stdout := itest.TelepresenceOk(ctx, "intercept", "--namespace", s.AppNamespace(), "--mount", "false", svc, "--port", strconv.Itoa(svcPort))
-			require.Contains(stdout, "Using StatefulSet echo-headless")
-			s.CapturePodLogs(ctx, "service=echo-headless", "traffic-agent", s.AppNamespace())
+		30*time.Second, // waitFor
+		3*time.Second,  // polling interval
+		`intercepted workload never show up in list`)
 
-			defer func() {
-				itest.TelepresenceOk(ctx, "leave", "echo-headless-"+s.AppNamespace())
-				if test.webhook {
-					require.NoError(dropWebhookAnnotation(ctx, "statefulset", "echo-headless", s.AppNamespace()))
-				}
-
-				// Switch to default user and uninstall the agent
-				itest.TelepresenceQuitOk(ctx)
-				dfltCtx := itest.WithUser(ctx, "default")
-				itest.TelepresenceOk(dfltCtx, "uninstall", "--agent", "echo-headless", "-n", s.AppNamespace())
-				itest.TelepresenceQuitOk(dfltCtx)
-				itest.TelepresenceOk(ctx, "connect", "--manager-namespace", s.ManagerNamespace())
-
-				require.Eventually(
-					func() bool {
-						stdout, _, err := itest.Telepresence(ctx, "list", "--namespace", s.AppNamespace(), "--agents")
-						return err == nil && !strings.Contains(stdout, "echo-headless")
-					},
-					30*time.Second, // waitFor
-					3*time.Second,  // polling interval
-					`agent is never removed`)
-			}()
-
-			require.Eventually(
-				func() bool {
-					stdout, _, err := itest.Telepresence(ctx, "list", "--namespace", s.AppNamespace(), "--intercepts")
-					return err == nil && strings.Contains(stdout, "echo-headless: intercepted")
-				},
-				30*time.Second, // waitFor
-				3*time.Second,  // polling interval
-				`intercepted workload never show up in list`)
-
-			itest.PingInterceptedEchoServer(ctx, svc, "8080")
-		})
-	}
-}
-
-func annotateForWebhook(ctx context.Context, objKind, objName, objNamespace string, servicePort int) error {
-	err := itest.Kubectl(ctx, objNamespace, "patch", objKind, objName, "-p", fmt.Sprintf(`
-{
-	"spec": {
-		"template": {
-			"metadata": {
-				"annotations": {
-					"telepresence.getambassador.io/inject-traffic-agent": "enabled",
-					"telepresence.getambassador.io/inject-service-port": "%d"
-				}
-			}
-		}
-	}
-}`, servicePort))
-	if err != nil {
-		return err
-	}
-	return itest.RolloutStatusWait(ctx, objNamespace, objKind+"/"+objName)
-}
-
-func dropWebhookAnnotation(ctx context.Context, objKind, objName, objNamespace string) error {
-	err := itest.Kubectl(ctx, objNamespace, "patch", objKind, objName, "--type=json", "-p", `[{
-	"op": "remove",
-	"path": "/spec/template/metadata/annotations/telepresence.getambassador.io~1inject-traffic-agent"
-},
-{
-	"op": "remove",
-	"path": "/spec/template/metadata/annotations/telepresence.getambassador.io~1inject-service-port"
-}
-]`)
-	if err != nil {
-		return err
-	}
-	return itest.RolloutStatusWait(ctx, objNamespace, objKind+"/"+objName)
+	itest.PingInterceptedEchoServer(ctx, svc, "8080")
 }

--- a/integration_test/intercept_env_test.go
+++ b/integration_test/intercept_env_test.go
@@ -36,6 +36,7 @@ func (s *interceptEnvSuite) Test_ExcludeVariables() {
 	defer os.RemoveAll("echo.env") //nolint:errcheck // dont need to catch the err
 
 	// when
+	itest.TelepresenceOk(ctx, "connect", "--manager-namespace", s.ManagerNamespace())
 	itest.TelepresenceOk(ctx, "intercept", "echo-easy", "--namespace", s.AppNamespace(), "--env-file", "echo.env")
 
 	// then

--- a/integration_test/intercept_mount_test.go
+++ b/integration_test/intercept_mount_test.go
@@ -7,6 +7,7 @@ import (
 	"net/http"
 	"os"
 	"path/filepath"
+	"regexp"
 	"runtime"
 	goRuntime "runtime"
 	"strconv"
@@ -83,8 +84,10 @@ func (s *interceptMountSuite) Test_InterceptMount() {
 	require := s.Require()
 	ctx := s.Context()
 
-	stdout := itest.TelepresenceOk(ctx, "--namespace", s.AppNamespace(), "list", "--intercepts")
-	s.Regexp(s.ServiceName()+`\s*: intercepted`, stdout)
+	s.Eventually(func() bool {
+		stdout, _, err := itest.Telepresence(ctx, "list", "--namespace", s.AppNamespace(), "--intercepts")
+		return err == nil && regexp.MustCompile(s.ServiceName()+`\s*: intercepted`).MatchString(stdout)
+	}, 10*time.Second, time.Second)
 
 	time.Sleep(200 * time.Millisecond) // List is really fast now, so give the mount some time to become effective
 	st, err := os.Stat(s.mountPoint)
@@ -118,8 +121,10 @@ func (s *singleServiceSuite) Test_InterceptMountRelative() {
 	}()
 	s.Contains(stdout, "Using Deployment "+s.ServiceName())
 
-	stdout = itest.TelepresenceOk(ctx, "--namespace", s.AppNamespace(), "list", "--intercepts")
-	s.Regexp(s.ServiceName()+`\s*: intercepted`, stdout)
+	s.Eventually(func() bool {
+		stdout, _, err := itest.Telepresence(ctx, "list", "--namespace", s.AppNamespace(), "--intercepts")
+		return err == nil && regexp.MustCompile(s.ServiceName()+`\s*: intercepted`).MatchString(stdout)
+	}, 10*time.Second, time.Second)
 
 	time.Sleep(200 * time.Millisecond) // List is really fast now, so give the mount some time to become effective
 	mountPoint := filepath.Join(nwd, "rel-dir")
@@ -183,8 +188,10 @@ func (s *singleServiceSuite) Test_NoInterceptorResponse() {
 		itest.TelepresenceOk(ctx, "leave", fmt.Sprintf("%s-%s", s.ServiceName(), s.AppNamespace()))
 	}()
 	s.Contains(stdout, "Using Deployment "+s.ServiceName())
-	stdout = itest.TelepresenceOk(ctx, "--namespace", s.AppNamespace(), "list", "--intercepts")
-	s.Regexp(s.ServiceName()+`\s*: intercepted`, stdout)
+	s.Eventually(func() bool {
+		stdout, _, err := itest.Telepresence(ctx, "list", "--namespace", s.AppNamespace(), "--intercepts")
+		return err == nil && regexp.MustCompile(s.ServiceName()+`\s*: intercepted`).MatchString(stdout)
+	}, 10*time.Second, time.Second)
 
 	time.Sleep(2000 * time.Millisecond) // List is really fast now, so give the mount some time to become effective
 	s.CapturePodLogs(ctx, "app="+s.ServiceName(), "traffic-agent", s.AppNamespace())

--- a/integration_test/itest/cluster.go
+++ b/integration_test/itest/cluster.go
@@ -725,7 +725,12 @@ func TelepresenceOk(ctx context.Context, args ...string) string {
 	stdout, stderr, err := Telepresence(ctx, args...)
 	assert.NoError(t, err, "telepresence was unable to run, stdout %s", stdout)
 	if err == nil {
-		assert.Empty(t, stderr, "Expected stderr to be empty, but got: %s", stderr)
+		if strings.HasPrefix(stderr, "Warning:") && !strings.ContainsRune(stderr, '\n') {
+			// Accept warnings, but log them.
+			dlog.Warn(ctx, stderr)
+		} else {
+			assert.Empty(t, stderr, "Expected stderr to be empty, but got: %s", stderr)
+		}
 	}
 	return stdout
 }

--- a/integration_test/large_files_test.go
+++ b/integration_test/large_files_test.go
@@ -49,7 +49,7 @@ const (
 )
 
 func init() {
-	itest.AddConnectedSuite("", func(h itest.NamespacePair) itest.TestingSuite {
+	itest.AddTrafficManagerSuite("", func(h itest.NamespacePair) itest.TestingSuite {
 		return &largeFilesSuite{
 			Suite:         itest.Suite{Harness: h},
 			NamespacePair: h,
@@ -81,28 +81,29 @@ func (s *largeFilesSuite) SetupSuite() {
 	s.Suite.SetupSuite()
 	ctx := s.Context()
 	require := s.Require()
-	require.NoError(s.TelepresenceHelmInstall(ctx, true, "--set", "timeouts.agentArrival=90s"))
-	itest.TelepresenceOk(ctx, "connect", "--manager-namespace", s.ManagerNamespace())
 	wg := sync.WaitGroup{}
 	wg.Add(s.ServiceCount())
 	k8s := filepath.Join("testdata", "k8s")
 	for i := 0; i < s.ServiceCount(); i++ {
 		go func(i int) {
 			defer wg.Done()
+			svc := fmt.Sprintf("%s-%d", s.Name(), i)
 			rdr, err := itest.OpenTemplate(ctx, filepath.Join(k8s, "hello-pv-volume.yaml"), &qname{
-				Name:       fmt.Sprintf("%s-%d", s.Name(), i),
+				Name:       svc,
 				Namespace:  s.AppNamespace(),
 				VolumeSize: "350Mi", // must cover fileCountPerSvc * fileSize
 			})
 			require.NoError(err)
 			require.NoError(s.Kubectl(dos.WithStdin(ctx, rdr), "apply", "-f", "-"))
+			s.NoError(itest.RolloutStatusWait(ctx, s.AppNamespace(), "deploy/"+svc))
 		}(i)
 	}
 	wg.Wait()
 }
 
 func (s *largeFilesSuite) createIntercepts(ctx context.Context) {
-	require := s.Require()
+	itest.TelepresenceOk(ctx, "connect", "--manager-namespace", s.ManagerNamespace())
+
 	wg := sync.WaitGroup{}
 	wg.Add(s.ServiceCount())
 	for i := 0; i < s.ServiceCount(); i++ {
@@ -117,6 +118,7 @@ func (s *largeFilesSuite) createIntercepts(ctx context.Context) {
 				svc,
 			)
 			var info intercept.Info
+			require := s.Require()
 			require.NoError(json.Unmarshal([]byte(stdout), &info))
 			require.Equal(fmt.Sprintf("%s-%s", svc, s.AppNamespace()), info.Name, ioutil.WriterToString(info.WriteTo))
 			require.NotNil(info.Mount)
@@ -146,6 +148,7 @@ func (s *largeFilesSuite) TearDownSuite() {
 			s.NoError(s.Kubectl(dos.WithStdin(ctx, rdr), "delete", "-f", "-"))
 		}(i)
 	}
+	itest.TelepresenceQuitOk(ctx)
 	wg.Wait()
 }
 
@@ -215,6 +218,7 @@ func (s *largeFilesSuite) largeFileIntercepts(ctx context.Context) {
 		s.T().FailNow()
 	}
 	s.createIntercepts(ctx)
+	defer s.leaveIntercepts(ctx)
 
 	// Validate the first entry
 	wg.Add(s.ServiceCount())

--- a/integration_test/multiple_intercepts_test.go
+++ b/integration_test/multiple_intercepts_test.go
@@ -6,6 +6,7 @@ import (
 	"io"
 	"net"
 	"net/http"
+	"regexp"
 	"strconv"
 	"sync"
 	"time"
@@ -71,10 +72,18 @@ func (s *multipleInterceptsSuite) TearDownSuite() {
 
 func (s *multipleInterceptsSuite) Test_Intercepts() {
 	ctx := s.Context()
-	stdout := itest.TelepresenceOk(ctx, "--namespace", s.AppNamespace(), "list", "--intercepts")
-	for i := 0; i < s.ServiceCount(); i++ {
-		s.Regexp(fmt.Sprintf(`%s-%d\s*: intercepted`, s.Name(), i), stdout)
-	}
+	s.Eventually(func() bool {
+		stdout, _, err := itest.Telepresence(ctx, "list", "--namespace", s.AppNamespace(), "--intercepts")
+		if err != nil {
+			return false
+		}
+		for i := 0; i < s.ServiceCount(); i++ {
+			if !regexp.MustCompile(fmt.Sprintf(`%s-%d\s*: intercepted`, s.Name(), i)).MatchString(stdout) {
+				return false
+			}
+		}
+		return true
+	}, 10*time.Second, time.Second)
 
 	wg := sync.WaitGroup{}
 	wg.Add(s.ServiceCount())

--- a/integration_test/to_pod_test.go
+++ b/integration_test/to_pod_test.go
@@ -3,6 +3,7 @@ package integration_test
 import (
 	"fmt"
 	"net"
+	"regexp"
 	"sync"
 	"time"
 
@@ -19,8 +20,10 @@ func (s *connectedSuite) Test_ToPodPortForwarding() {
 	stdout := itest.TelepresenceOk(ctx, "intercept", "--namespace", s.AppNamespace(), "--mount", "false", svc, "--port", "8080", "--to-pod", "8081", "--to-pod", "8082")
 	defer itest.TelepresenceOk(ctx, "leave", svc+"-"+s.AppNamespace())
 	require.Contains(stdout, "Using Deployment "+svc)
-	stdout = itest.TelepresenceOk(ctx, "list", "--namespace", s.AppNamespace(), "--intercepts")
-	require.Contains(stdout, svc+": intercepted")
+	s.Eventually(func() bool {
+		stdout, _, err := itest.Telepresence(ctx, "list", "--namespace", s.AppNamespace(), "--intercepts")
+		return err == nil && regexp.MustCompile(svc+`\s*: intercepted`).MatchString(stdout)
+	}, 10*time.Second, time.Second)
 
 	wg := sync.WaitGroup{}
 	wg.Add(1)

--- a/pkg/client/cli/cloud/message.go
+++ b/pkg/client/cli/cloud/message.go
@@ -15,6 +15,7 @@ import (
 	"google.golang.org/grpc/credentials"
 	"google.golang.org/protobuf/types/known/emptypb"
 
+	"github.com/datawire/dlib/dlog"
 	"github.com/datawire/dlib/dtime"
 	"github.com/telepresenceio/telepresence/rpc/v2/systema"
 	"github.com/telepresenceio/telepresence/v2/pkg/client"
@@ -31,18 +32,18 @@ type messageCache struct {
 }
 
 // newMessageCache returns a new messageCache, initialized from the users' if it exists.
-func newMessageCache(ctx context.Context) (*messageCache, error) {
+func newMessageCache(ctx context.Context) *messageCache {
 	cmc := &messageCache{}
 
 	if err := cache.LoadFromUserCache(ctx, cmc, messagesCacheFilename); err != nil {
 		if !os.IsNotExist(err) {
-			return nil, err
+			dlog.Warn(ctx, err)
 		}
 		cmc.NextCheck = time.Time{}
 		cmc.MessagesDelivered = make(map[string]struct{})
 		_ = cache.SaveToUserCache(ctx, cmc, messagesCacheFilename)
 	}
-	return cmc, nil
+	return cmc
 }
 
 // getMessages communicates with Ambassador Cloud and stores those messages
@@ -112,10 +113,7 @@ func RaiseMessage(cmd *cobra.Command, _ []string) error {
 	cmdUsed := strings.Split(cmd.Use, " ")[0]
 
 	// Load the config
-	cmc, err := newMessageCache(ctx)
-	if err != nil {
-		return err
-	}
+	cmc := newMessageCache(ctx)
 
 	// Check if it is time to get new messages from Ambassador Cloud
 	if dtime.Now().After(cmc.NextCheck) {

--- a/pkg/client/cli/cloud/messages_test.go
+++ b/pkg/client/cli/cloud/messages_test.go
@@ -34,12 +34,9 @@ func Test_GetMessageFromCache(t *testing.T) {
 	ctx := newTestContext(t)
 
 	// Pre-load cmc with a message for intercept
-	cmc, err := newMessageCache(ctx)
+	cmc := newMessageCache(ctx)
 	ceptMessage := "Test Intercept Message"
 	cmc.Intercept = ceptMessage
-	if err != nil {
-		t.Error(err)
-	}
 
 	tests := []struct {
 		name string
@@ -82,12 +79,9 @@ func Test_UpdateMessages(t *testing.T) {
 	ctx := newTestContext(t)
 
 	// Pre-load cmc with a message for intercept
-	cmc, err := newMessageCache(ctx)
+	cmc := newMessageCache(ctx)
 	ceptMessage := "Test Old Intercept Message"
 	cmc.Intercept = ceptMessage
-	if err != nil {
-		t.Error(err)
-	}
 
 	// Ensure we get the current message in the cache
 	res := cmc.getMessageFromCache(ctx, "intercept")
@@ -142,10 +136,7 @@ func Test_RefreshMessagesConfig(t *testing.T) {
 		t.Error(err)
 	}
 
-	cmc, err := newMessageCache(ctx)
-	if err != nil {
-		t.Error(err)
-	}
+	cmc := newMessageCache(ctx)
 	// Mock what we get from `GetUnauthenticatedCommandMessages`
 	ceptMessage := "Intercept message, testing config"
 	updatedMessageResponse := &systema.CommandMessageResponse{

--- a/pkg/tunnel/pipe.go
+++ b/pkg/tunnel/pipe.go
@@ -44,7 +44,7 @@ func (s channelStream) ID() ConnID {
 func (s channelStream) Receive(ctx context.Context) (Message, error) {
 	select {
 	case <-ctx.Done():
-		return nil, nil
+		return nil, ctx.Err()
 	case m, ok := <-s.recvCh:
 		if !ok {
 			return nil, io.EOF

--- a/pkg/tunnel/stream.go
+++ b/pkg/tunnel/stream.go
@@ -105,15 +105,13 @@ func ReadLoop(ctx context.Context, s Stream) (<-chan Message, <-chan error) {
 				endReason = "EOF on input"
 			case errors.Is(err, net.ErrClosed):
 				endReason = "stream closed"
+			case errors.Is(err, context.Canceled), status.Code(err) == codes.Canceled:
+				endReason = err.Error()
 			default:
-				if status.Code(err) == codes.Canceled {
-					endReason = "stream closed"
-				} else {
-					endReason = err.Error()
-					select {
-					case errCh <- fmt.Errorf("!! %s %s, read from grpc.ClientStream failed: %w", s.Tag(), s.ID(), err):
-					default:
-					}
+				endReason = err.Error()
+				select {
+				case errCh <- fmt.Errorf("!! %s %s, read from grpc.ClientStream failed: %w", s.Tag(), s.ID(), err):
+				default:
 				}
 			}
 			break

